src/wayland/wl.c

```diff
@@ -0,0 +1,743 @@
+                        version >= ORG_KDE_KWIN_IDLE_TIMEOUT_IDLE_SINCE_VERSION) {
+                ctx.idle_handler = wl_registry_bind(registry, name, &org_kde_kwin_idle_interface, 1);
+        }
+}
+
+static void handle_global_remove(void *data, struct wl_registry *registry,
+                uint32_t name) {
+        struct dunst_output *output, *tmp;
+        wl_list_for_each_safe(output, tmp, &ctx.outputs, link) {
+                if (output->global_name == name) {
+                        destroy_output(output);
+                        break;
+                }
+        }
+}
+
+static const struct wl_registry_listener registry_listener = {
+        .global = handle_global,
+        .global_remove = handle_global_remove,
+};
+
+bool init_wayland() {
+        wl_list_init(&ctx.outputs);
+        //wl_list_init(&ctx.seats);
+
+        ctx.display = wl_display_connect(NULL);
+
+        if (ctx.display == NULL) {
+                LOG_E("failed to create display");
+                return false;
+        }
+
+        ctx.registry = wl_display_get_registry(ctx.display);
+        wl_registry_add_listener(ctx.registry, &registry_listener, NULL);
+        wl_display_roundtrip(ctx.display);
+
+        if (ctx.compositor == NULL) {
+                LOG_E("compositor doesn't support wl_compositor");
+                return false;
+        }
+        if (ctx.shm == NULL) {
+                LOG_E("compositor doesn't support wl_shm");
+                return false;
+        }
+        if (ctx.layer_shell == NULL) {
+                LOG_E("compositor doesn't support zwlr_layer_shell_v1");
+                return false;
+        }
+        if (ctx.seat == NULL) {
+                LOG_W("no seat was found, so dunst cannot see input");
+        }
+        else {
+                // pointer may not be detected yet
+                /* if (ctx.pointer.wl_pointer == NULL) { */
+                        /* LOG_W("no pointer was found, so dunst cannot see input"); */
+                /* } */
+                if (ctx.idle_handler == NULL) {
+                        LOG_I("compositor doesn't support org_kde_kwin_idle_interface");
+                }
+                else if (ctx.idle_timeout == NULL) {
+                        // something went wrong in setting the timeout
+                        LOG_W("couldn't set idle timeout");
+                }
+        }
+
+        if (ctx.xdg_output_manager != NULL) {
+                struct dunst_output *output;
+                wl_list_for_each(output, &ctx.outputs, link) {
+                        get_xdg_output(output);
+                }
+                wl_display_roundtrip(ctx.display);
+        }
+        //if (ctx.xdg_output_manager == NULL &&
+        //                strcmp(ctx.config.output, "") != 0) {
+        //        LOG_E("warning: configured an output but compositor doesn't "
+        //                "support xdg-output-unstable-v1 version 2");
+        //}
```

Comment:
Why is this commented out?

Reply:
It didn't compile, and I don't know how to fix that check correctly.

Looking at it again, it is a warning for when you have configured a preferred monitor name and you compositor doesn't support the protocol.
I don't think that it is relevant to us, because we use output numbers. I'll test with a second monitor soon, to see if `monitor` works without xdg-output-unstable-v1.

Reply:
As I thought, we don't need xdg-output. This warning can be safely deleted. Actually all the xdg output code can be removed. If we want to have an option to select monitor names to dunst, we will need to add it again. Is it worth to remove all xdg-output code then?

Reply:
Since we use numbers to define monitors in the config currently, I think we can safely remove it. Less is more as they say.

Reply:
I removed all the code regarding xdg-output. That was quite a lot of code for just getting the window name...